apex/contrib/optimizers/distributed_fused_adam.py

```diff
@@ -1050,9 +1055,22 @@ def clip_grad_norm(self, max_norm, parameters=None, norm_type=2.0):
         assert max_norm > 0
         total_norm = self.grad_norm(parameters=parameters, norm_type=norm_type)
         clip_coef = max_norm / (total_norm + 1e-6)
-        self._grad_scale = torch.minimum(self._grad_scale, clip_coef)
+        clip_coef_clamped = torch.clamp(clip_coef, max=1.0)
+        self._grad_scale *= clip_coef_clamped
         return total_norm
 
+    def unscale_grads(self, inv_scale, *args):
+        """Custom unscale function for use by AMP gradient scaler
+
+        Overflow checking is deferred to optimization step.
+
+        Arguments:
+            inv_scale (torch.Tensor): factor to multiply gradients
+
+        """
+        self._grad_scale *= inv_scale.view([])
```

Comment:
qq: why `.view`?

Reply:
The grad scaler initializes its scale factor as a 1D tensor of size 1 instead of as a scalar: https://github.com/pytorch/pytorch/blob/6e1764d806bc45e7c15c79f7e5f1a0bafb76ec73/torch/cuda/amp/grad_scaler.py#L146